### ISSUE TITLE
Is there an async option while calling predict in the client?

### ISSUE THREAD

**Issue by balajivenkatesh** [replicate/replicate-python#107]
from quick look, it seems like [this line](https://github.com/replicate/replicate-python/blob/4d4a6ddd692a87272082933f780e9a0ae6cf5fd7/replicate/version.py#L39) makes it blocking. 

Is there a way to work around this?

**Comment by peter65374** [comment]
I guess we have to implement async mode by aiohttp/httpx by ourselves. And I am worried about the concurrent API connection limit of replicate server.

**Comment by antont** [comment]
Yep also this is blocking in client's `handle_async_request`

https://github.com/replicate/replicate-python/blob/cc1c767803a63af7eeaf6e2acc411acc3b7b642d/replicate/client.py#L264

Ah apparently it's coming, discussion in https://github.com/replicate/replicate-python/pull/76#issuecomment-1738827381

**Comment by jp-hedilog** [comment]
@peter65374 were u able to figure out what the concurrent API connection limit is?

**Comment by peter65374** [comment]
@jp-hedilog I wrote my own async functions by aiohttp instread of using replicate-python sdk. e.g. Call replicate HTTP Restful API to check prediction status as below.
```python
async def get_prediction_status(self, prediction_id: str)->Tuple[Union[str, None], str]:
        url = f"https://api.replicate.com/v1/predictions/{prediction_id}"
        self.logger.info(f"get prediction status...")
        headers={
            "Content-Type": "application/json",
            "Accept": "application/json",
            "Authorization": f"Token {self.replicate_api_key}"
        }
        prediction_output = None
        status = "starting"
        start = time.perf_counter()
        async with aiohttp.ClientSession() as session:
            async with session.get(url, headers=headers) as response:
                self.logger.info(f"get_prediction status response code: {response.status}")
                if response.status != 200:
                    self.logger.error(f"Non-200 response: {response.status}")
                    raise Exception("Non-200 response: " + str(await response.text()))
                resp_json = await response.json()
                self.logger.debug(f"replicate get prediction status response: {resp_json}")
                status = resp_json["status"]
                prediction_output = resp_json.get("output", None)
                if status == 'failed':
                    self.logger.error(f"Prediction failed:\n {resp_json['logs']}")
                    prediction_output = resp_json.get("error")
                self.logger.info(f"status: {status}, output: {prediction_output}")
        duration = time.perf_counter() - start
        self.logger.info(f"Total time taken to get prediction status by replicate: {duration:.2f} seconds.")
        return prediction_output, status
```

**Comment by mattt** [comment]
Hi, everyone. I'm happy to report that async support just landed in version [0.18.0](https://github.com/replicate/replicate-python/releases/tag/0.18.0). Please give it a try and let me know how it works for you.

> And I guess QPS limit == 1 for public model users, you could raise limit by new 'Deployment' feature provided by Replicate.com, I guess.

@peter65374 Public models on Replicate aren't rate-limited like this. Here's an example of how to run multiple predictions from the public SDXL model at the same time (it should complete in roughly the same amount of time as the longest-running predictions, which at time of writing is ~12 seconds):

```python
import asyncio
import replicate

# https://replicate.com/stability-ai/sdxl
model_version = "stability-ai/sdxl:39ed52f2a78e934b3ba6e2a89f5b1c712de7dfea535525255b1aa35c5565e08b"
prompts = [
    f"A chariot pulled by a team of {count} rainbow unicorns"
    for count in ["two", "four", "six", "eight"]
]

async with asyncio.TaskGroup() as tg:
    tasks = [
        tg.create_task(replicate.async_run(model_version, input={"prompt": prompt}))
        for prompt in prompts
    ]

results = await asyncio.gather(*tasks)
print(results)
```
